extensions/pdh/tests/PerformanceDataMonitorTests.cpp

```diff
@@ -0,0 +1,268 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <memory>
+#include <string>
+#include <vector>
+#include <set>
+#include <fstream>
+
+#include "TestBase.h"
+#include "processors/PutFile.h"
+#include "utils/file/FileUtils.h"
+#include "PerformanceDataMonitor.h"
+#include "rapidjson/filereadstream.h"
+
+using PutFile = org::apache::nifi::minifi::processors::PutFile;
+using PerformanceDataMonitor = org::apache::nifi::minifi::processors::PerformanceDataMonitor;
+using PerformanceDataCounter = org::apache::nifi::minifi::processors::PerformanceDataCounter;
+
+class PerformanceDataMonitorTester {
+ public:
+  PerformanceDataMonitorTester() {
+    LogTestController::getInstance().setTrace<TestPlan>();
+    dir_ = test_controller_.createTempDirectory("/tmp/gt.XXXXXX");
+    plan_ = test_controller_.createPlan();
+    performance_monitor_ = plan_->addProcessor("PerformanceDataMonitor", "pdhsys");
+    putfile_ = plan_->addProcessor("PutFile", "putfile", core::Relationship("success", "description"), true);
+    plan_->setProperty(putfile_, PutFile::Directory.getName(), dir_);
+  }
+
+  void runProcessors() {
+    plan_->runNextProcessor();      // PerformanceMonitor
+    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+    plan_->runCurrentProcessor();   // PerformanceMonitor
+    plan_->runNextProcessor();      // PutFile
+    plan_->runCurrentProcessor();   // PutFile
+  }
+
+  void setPerformanceMonitorProperty(core::Property property, const std::string& value) {
+    plan_->setProperty(performance_monitor_, property.getName(), value);
+  }
+
+  TestController test_controller_;
+  std::string dir_;
+  std::shared_ptr<TestPlan> plan_;
+  std::shared_ptr<core::Processor> performance_monitor_;
+  std::shared_ptr<core::Processor> putfile_;
+};
+
+
+TEST_CASE("PerformanceDataMonitorEmptyPropertiesTest", "[performancedatamonitoremptypropertiestest]") {
+  PerformanceDataMonitorTester tester;
+  tester.runProcessors();
+
+  REQUIRE(tester.test_controller_.getLog().getInstance().contains("No valid counters for PerformanceDataMonitor", std::chrono::seconds(0)));
+
+  uint32_t number_of_flowfiles = 0;
+  auto lambda = [&number_of_flowfiles](const std::string& path, const std::string& filename) -> bool {
+    ++number_of_flowfiles;
+    return true;
+  };
+
+  utils::file::FileUtils::list_dir(tester.dir_, lambda, tester.plan_->getLogger(), false);
```

Comment:
there is a `list_dir_all` returning a `vector`, that could be an alternative here, not requiring a lambda

Reply:
changed in [d5959c3](https://github.com/martinzink/nifi-minifi-cpp/commit/d5959c3b1b5d0f377e7b7f2c13cc22a484b2e4a6)